### PR TITLE
Neon extension

### PR DESCRIPTION
**Pull request by sergiusens**
- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/project_loader/_extensions/_utils.py [hunk 1]
```diff
@@ -98,12 +98,9 @@ def find_extension(extension_name: str) -> Type[Extension]:
...
 
-    # This may throw an AttributeError, but that would be programmer error of whoever
-    # is hacking on extensions.
-    extension_class_name = "{}Extension".format(
-        extension_name.capitalize().replace("-", "_")
-    )
-    return getattr(extension_module, extension_class_name)
+    # Extensions have an attribute with the name of the extension
+    # class.
+    return getattr(extension_module, "ExtensionImpl")
```
Comment (cjp256) [inline snapcraft/internal/project_loader/_extensions/_utils.py]: Seems a little odd to require them all to share the same name, but it works and simplifies the naming rule.

File: snapcraft/internal/project_loader/_extensions/kde_neon.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    def __init__(self, *, extension_name: str, yaml_data: Dict[str, Any]) -> None:
+        super().__init__(extension_name=extension_name, yaml_data=yaml_data)
+
+        platform_snap = _PLATFORM_SNAP[yaml_data.get("base")]
+        self.root_snippet = {
+            "plugs": {
+                "gtk-3-themes": {
+                    "interface": "content",
+                    "target": "$SNAP/data-dir/themes",
+                    "default-provider": "gtk-common-themes:gtk-3-themes",
```
Comment (cjp256) [inline snapcraft/internal/project_loader/_extensions/kde_neon.py]: I assume these are shared between gtk and kde?

Comment (sergiusens) [reply snapcraft/internal/project_loader/_extensions/kde_neon.py]: I am not totally sold on this, just put it out there, it was a proposal from @kenvandine and part of the original grand scheme.

File: snapcraft/internal/project_loader/_extensions/_utils.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/internal/project_loader/_extensions/_utils.py]: As a follow up to your point on IRC.  I think the naming scheme is fine.  As implemented, it's already a requirement on the module name, with the added tweak to capitalize the class name.

I kinda liked how the plugin system found it by class type, but that's overkill and has some potential issues of its own.

I do often use class names to navigate source at times, so I'd lean to towards the naming format unless there's something technically wrong with it.  However, as it's largely a personal preference, I defer to your judgement.   :)

**Comment by sergiusens** [comment]
Thanks @apachelogger I will take a look at the locale-gen bits.
We can certainly add usr/sbin and use/games.
And I will make sure I find where vlc's pulse plugins live inside the content snap and add those.

I guess this would benefit GUI users more as is now in almost all cases, we can work on another  neon-cli extension for that specific case, are there apps around that are in need of this?

**Comment by apachelogger** [comment]
FWIW I'm not objecting to landing as-is, merely pointing out the deficiencies.

usr/games is used by some of the kde snaps because they internally call some helper binaries we get from staged debs which use usr/games for some reason. The vlc plugin needs to be found for output to actually work which impacts also some kde games, `kblinken` specifically comes to mind. In both cases @jriddell would know better how widely used these things are right now.
locale-gen is required for a feature that is in just about every KDE application (Help -> Switch application language)

**Comment by sergiusens** [comment]
Thanks @apachelogger, I added locale-gen functionality (missing the one feature which you apparently want so I will work on that now, I wanted to understand it better), vlc plugin paths and the paths for games

**Comment by cjp256** [comment]
The one thought I did have, is the environment info for SNAP_DESKTOP_RUNTIME - is that an interface?  Or would it perhaps be better if that configuration was provided by the runtime?

**Comment by sergiusens** [comment]
on $SNAP_DESKTOP_RUNTIME, I thought about it, and was going for it initially, but that means that most of the common code related to snaps gets split out into all the _frameworks_, this would also require a wrapper and not be able to be offered as a command-chain as it would be an out-of-snap runnable for which we might get away with by using a wrapper and by doing so add one more level of indirection.

**Pull request merged**